echopype/clean/impulse_noise.py

```diff
@@ -0,0 +1,463 @@
+        maskf = np.ma.masked_greater(forward, thr).mask
+        maskb = np.ma.masked_greater(backward, thr).mask
+        mask = mask | (maskf & maskb)
+
+    # get second mask indicating valid samples in IN mask
+    mask_[:, 0 : max(n)] = True
+    mask_[:, -max(n) :] = True
+
+    return mask, mask_
+
+
+def _wang(
+    Sv: np.ndarray,
+    thr: Tuple[float, float],
+    erode: List[Tuple[int, int]],
+    dilate: List[Tuple[int, int]],
+    median: List[Tuple[int, int]],
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Clean impulse noise from Sv data following the method described by:
+
+        Wang et al. (2015) ’A noise removal algorithm for acoustic data with
+        strong interference based on post-processing techniques’, CCAMLR
+        SG-ASAM: 15/02.
+
+    This algorithm runs different cycles of erosion, dilation, and median
+    filtering to clean impulse noise from Sv. Note that this function
+    returns a clean/corrected Sv array, instead of a boolean array indicating
+    the occurrence of impulse noise.
+
+    Parameters
+    ----------
+        Sv     (float)    : 2D numpy array with Sv data (dB).
+        thr    (int/float): 2-element tuple with bottom/top Sv thresholds (dB)
+        erode  (int)      : list of 2-element tuples indicating the window's
+                            size for each erosion cycle.
+        dilate (int)      : list of 2-element tuples indicating the window's
+                            size for each dilation cycle.
+        median (int)      : list of 2-element tuples indicating the window's
+                            size for each median filter cycle.
+
+    Returns
+    -------
+        float             : 2D array with clean Sv data.
+        bool              : 2D array with mask indicating valid clean Sv data.
+    """
+
+    # set weak noise and strong interference as vacant samples (-999)
+    Sv_thresholded = Sv.copy()
+    Sv_thresholded[(Sv < thr[0]) | (Sv > thr[1])] = -999
+
+    # remaining weak interferences will take neighbouring vacant values
+    # by running erosion cycles
+    Sv_eroded = Sv.copy()
+    for e in erode:
+        Sv_eroded = erosion(Sv_thresholded, np.ones(e))
+
+    # the last step might have turned interferences inside biology into vacant
+    # samples, this is solved by running dilation cycles
+    Sv_dilated = Sv_eroded.copy()
+    for d in dilate:
+        Sv_dilated = dilation(Sv_dilated, np.ones(d))
+
+    # dilation has modified the Sv value of biological features, so these are
+    # now corrected to corresponding Sv values before the erosion/dilation
+    Sv_corrected1 = Sv_dilated.copy()
+    mask_bio = (Sv_dilated >= thr[0]) & (Sv_dilated < thr[1])
+    Sv_corrected1[mask_bio] = Sv_thresholded[mask_bio]
+
+    # compute median convolution in Sv corrected array
+    Sv_median = Sv_corrected1.copy()
+    for m in median:
+        Sv_median = mask_transformation.log(
+            medianf(mask_transformation.lin(Sv_median), footprint=np.ones(m))
```

Comment:
Looks like median filter is wrapped by dask already: https://image.dask.org/en/latest/dask_image.ndfilters.html
Could you check if lazy-loaded dask arrays are handled?

Reply:
The function itself appears to handle them, but I'm not sure using it actually makes any sense, since we have to drop to numpy in the code right before it anyway, due to grey_erosion/dilation not being implemented in dask_image.ndmorph